astyanax-recipes/src/main/java/com/netflix/astyanax/recipes/storage/ObjectReader.java

```diff
@@ -166,9 +174,15 @@ public void run() {
                     }
                     finally {
                         executor.shutdown();
-                        if (!executor.awaitTermination(maxWaitTimeInSeconds, TimeUnit.SECONDS)) {
-                            throw new Exception("Took too long to fetch object: " + objectName);
+                        while (!executor.isTerminated()) {
+                            try {
+                                if (!executor.awaitTermination(maxWaitTimeInSeconds, TimeUnit.SECONDS))
+                                    throw new Exception("Took too long to fetch object: " + objectName);
+                            } catch (InterruptedException e) {
+
+                            }
```

Comment:
Instead of the loop, you should probably just invoke executor.shutdownNow() if awaitTermination doesn't return or is itself interrupted, e.g. someone called shutdownNow() on ObjectReader's thread. Something like this:

                    finally {
                        executor.shutdown();
                        try {
                            if (!executor.awaitTermination(maxWaitTimeInSeconds, TimeUnit.SECONDS)) {
                                throw new Exception("Took too long to fetch object: " + objectName);
                            }
                        } finally{
                            if(!executor.isTerminated()) {
                                executor.shutdownNow();
                            }
                        }
                    }